File: src/main/scala/docscanner/ScanSource.scala
package docscanner

import scala.jdk.CollectionConverters.*
import scala.scalajs.js.timers.SetIntervalHandle
import scalajs.js.timers
import scalajs.js
import scalajs.js.Dynamic.global as g
import js.Dynamic.literal as l
import typings.node.bufferMod.global.Buffer
import typings.node.fsMod
import typings.node.anon.*
import typings.obsidian.mod
import typings.obsidian.mod.FileSystemAdapter
import utils.{Lexer, Utils}

import concurrent.ExecutionContext.Implicits.global
import scala.collection.mutable

//metauses: ["Lexer", "Utils"]
object ScanSource:
  private type DOC = String

  private val path = g.require("path")

  private var gitBranchName : String = _        // the working gitBranchToScan defined in the settings
  private var applicationPath : String = _      // application path
  private var codeExtension : String = _        // the source file name extension
  private var documentPath : String = _         // document path abs path
  private var relativeDocumentPath : String = _ // document path abs rel from vault root
  private var groupBySize : Int = _             // number of documents to process at a time
  //
  //
  private var sleepLength : Int = _             // number of seconds to sleep
  //
  // number of computation phases per iteration
  //
  private var phaseCount = 0                    // the number of the current phase
  //
  // current files with 'ext' extension with paths
  //
  private var applicationFileListWithExtension : List[List[String]] = _
  private var documentFileListWithExtension : List[String] = _

  private val sourceAndDocumentLink = mutable.Set[ DOC ]() // docs that have a source
  private val documentAndContentMap = mutable.HashMap[String, String]() // all documents created with the content

  /**#ScanSource
   * uses #Lexer #Utils
   * ## apply
   * At every sleep length millliseconds interval get all the business rules from the latest source file and write it to the document file. if the
   * source file changes then delete the doc file and get newest comments from the source file ^scanner-00
   *
   * 1. The 2 second interval must be made configurable. - Still to be implemented
   * 2. The following must be configurable
   *    1. Project file location
   *       2. Source file extension e.g. scala or java
   *       3. Relative path from the project file location to the start of the source
   *       4. Relative path of the extracted from the project path to the document location
   *
   * @param app
   * @param _appPath
   * @param _ext
   * @param _work
   * @param _docPath
   * @param _sleepLength
   * @return
   */
  def apply(app : mod.App, _appPath : String, _ext : String, _docPath : String, _sleepLength : Int, _groupBySize : Int, _gitBranchName : String): SetIntervalHandle =
    //
    // initialization
    //
    path.basename("")

    applicationPath = _appPath
    gitBranchName = _gitBranchName
    codeExtension = _ext
    documentPath = _docPath
    relativeDocumentPath = _docPath
    sleepLength = _sleepLength
    groupBySize = _groupBySize
    //
    // make sure the doc path relative from vault exist - if not create this path
    //
    val fsa = app.vault.adapter.asInstanceOf[FileSystemAdapter]
    val vaultPath = fsa.getBasePath()

    timers.setInterval(sleepLength)(this.run(fsa))

  /**
   * ## private def isBranchStillActive() : Boolean =
   * check if the current gitBranchToScan is still the active gitBranchToScan for the scanner
   * if the file does not exist then assume the user wants to use the default code in the application path
   * @return
   */
  private def isBranchStillActive: Boolean =
    if Utils.branchNameLocation.isEmpty then
      Utils.getBranchNameFileLocation(applicationPath)
    if Utils.branchNameLocation.isDefined then
      try {
        val branchName = fsMod.readFileSync(Utils.branchNameLocation.get + Utils.separator + "current-branch.txt", l(encoding = "utf8", flag = "r")
          .asInstanceOf[ObjectEncodingOptionsflagEncoding])
          .asInstanceOf[String]
        if branchName.isEmpty || !branchName.trim.equalsIgnoreCase(gitBranchName.trim) then
          false
        else
          true
      } catch {
        case ex  : js.JavaScriptException => true
      }
    else
      false

  private def run(fsa : FileSystemAdapter) : Unit =
    //
    // if the current gitBranchToScan is the defined gitBranchToScan then go ahead an process else do not process
    //
    if phaseCount == 0 then
      if !isBranchStillActive then return()

    //
    // do work in phases - get all the source files.
    //
    if phaseCount == 1 then
      applicationFileListWithExtension = Utils.walk(applicationPath
        .asInstanceOf[String])
        .filter(file => file.endsWith(codeExtension))
        .grouped(10)
        .toList
    //
    // get list of document files
    //
    if phaseCount == 2 then
      documentFileListWithExtension = Utils.listMDFilesInVault(fsa, documentPath)
    //
    // if the source if younger then the document file then
    //    load the lines from the source file and scan for comments.
    //    write comments out to document file
    //
    if phaseCount == 3 then
      if applicationFileListWithExtension.nonEmpty then
        applicationFileListWithExtension.head.foreach(srcFile =>
          val documentName = createDocNameFromSourceName(srcFile)
          val documentNameAndPath = s"$documentPath/$documentName"

          sourceAndDocumentLink += documentNameAndPath
          //
          // if the document file does not exist then create an empty one
          //
          val returnStat =  fsa.stat(documentNameAndPath).toFuture.foreach(stat =>
            val docStat = if stat != null then
              ( false, stat )
            else
              fsa.write(documentNameAndPath, "")
              ( true, null )

            try {
              val srcStat = fsMod.statSync(srcFile)

              val srcModTime = srcStat.get.mtimeMs

              if ( docStat._1 || docStat._2.mtime < srcModTime ) && isBranchStillActive then
                //
                // get comment lines from the srcFile then open doc file for writing and write the lines to it
                //
                val srcLines = fsMod.readFileSync(srcFile, l(encoding = "utf8", flag = "r")
                  .asInstanceOf[ObjectEncodingOptionsflagEncoding])
                  .asInstanceOf[String]

                val commentString = Lexer(srcLines)
                documentAndContentMap += ( s"$relativeDocumentPath${Utils.separator}$documentName".dropRight(3) -> commentString._1 )
                //
                // setup meta data in header of note
                //
                val metaData =
                  s"""---
                     |${commentString._2}---
                     |""".stripMargin

                fsa.write(documentNameAndPath, s"${metaData}[Source](file:$srcFile)\n\n---\n" + commentString._1).toFuture.foreach(Unit => ())
            } catch {
              case ex : js.JavaScriptException => println("source removed")
            }
          )
        )
        applicationFileListWithExtension = applicationFileListWithExtension.drop(1)
        phaseCount = 2

    if phaseCount == 4 && isBranchStillActive then
      //
      // every document that does not have a source file must be removed
      //
      documentFileListWithExtension.foreach((fileName: String) =>
        if !sourceAndDocumentLink.contains(fileName) then
          fsa.remove(fileName)
      )

      sourceAndDocumentLink.clear()
      documentAndContentMap.clear()
      phaseCount = -1

    phaseCount += 1

  /**
   * construct the document name given the source file name
   * @param sourceFile name used to construct the document name
   * @return the document name
   */
  private def createDocNameFromSourceName(sourceFile : String) =
    val docPath = s"${sourceFile.drop(applicationPath.length + 1).dropRight(codeExtension.length)}.md"
    val docName = docPath.replace(Utils.separator, ".")
    docName

File: ts/ScanSource.ts
import { App } from 'obsidian';
import SourceScanner from '../main';
import { statSync, readFileSync, existsSync, writeFileSync } from 'fs'
import { FileSystemAdapter } from 'obsidian'
import { Utils } from './Utils'
import { DocFolders } from './DocFolders';

export class ScanSource {

    codeScanner: (arg0: string) => string;

    applicationPath: string;          // application path
    codeExtension: string;            // the source file name extension
    documentPath: string;             // document path abs path
    testPath: string;                 // unit test code path 
    relativeDocumentPath: string;     // document path abs rel from vault root
    groupBySize: number;              // number of documents to process at a time 
    sleepLength: number;              // number of seconds to sleep
    phaseCount: number = 0;           // current processsing phase 

    applicationFileListWithExtension: Array<Array<string>>; // all application source files
    testFileListWithExtension: Array<Array<string>>;        // all test source files
    documentFileListWithExtension: Array<string>;           // all documents
    sourceAndDocumentLink = new Set<string>;

    fsa: FileSystemAdapter;
    utils: Utils;

    docFolders: DocFolders;            // document folders

    init(app: App, plugin: SourceScanner, scanner: (arg0: string) => string) {

        this.codeScanner = scanner;
        this.applicationPath = plugin.settings.applicationPath;
        this.codeExtension = plugin.settings.applicationExtension;
        this.documentPath = plugin.settings.documentPath;
        this.testPath = plugin.settings.unitTestPath;
        this.groupBySize = plugin.settings.groupBySize;
        this.sleepLength = plugin.settings.sleepLength;
        this.utils = new Utils(app);
        this.fsa = app.vault.adapter as FileSystemAdapter;

        return setInterval(() => this.run(), this.sleepLength);
    }

    run() {
        //
        // the folders might have been deleted or this is the first start of the app
        // so create the folders iff they do not exist
        //
        this.docFolders = this.utils.createFolders(this.documentPath);
        //
        // do work in phases
        // get all the implementation and test files into a list of chunks at most groupBySize
        //
        if (this.phaseCount == 1.0) {
            const impFiles = this.utils.filterFileNamesByExtension(
                this.codeExtension,
                this.utils.walkInFolderFromDir(this.applicationPath, []));

            this.applicationFileListWithExtension = [];

            for (let i = 0; i < impFiles.length; i += this.groupBySize) {
                const chunk = impFiles.slice(i, i + this.groupBySize);
                this.applicationFileListWithExtension.push(chunk);
            }

            const testFiles = this.utils.filterFileNamesByExtension(
                this.codeExtension,
                this.utils.walkInFolderFromDir(this.testPath, []));

            this.testFileListWithExtension = [];

            for (let i = 0; i < testFiles.length; i += this.groupBySize) {
                const chunk = testFiles.slice(i, i + this.groupBySize);
                this.testFileListWithExtension.push(chunk);
            }
        }
        //
        // get list of document files
        //        
        if (this.phaseCount == 2.0) {
            const files = this.utils.filterFileNamesByExtension(
                '.md',
                this.utils.walkInFolderFromDir(
                    this.fsa.getBasePath() + this.utils.separator + this.docFolders.settingsCommentFolder, []
                ));
            this.documentFileListWithExtension = files
                .map(fileName => {
                    return fileName.replace(this.fsa.getBasePath() + this.utils.separator, "");
                })
        }
        //
        // if the source if younger then the document file OR document file had to be created then
        //    load the lines from the source file and scan for comments.
        //    write comments out to document file
        // Make sure the files exist before getting their stat information
        //
        if (this.phaseCount == 3.0 && this.applicationFileListWithExtension.length > 0) {
            const filesToCheck = this.applicationFileListWithExtension.pop();
            if (filesToCheck != undefined) {
                filesToCheck.forEach(srcFile => {

                    const documentName = this.createDocNameFromSourceName(srcFile, this.applicationPath);
                    const documentNameAndPath =
                        `${this.documentPath}${this.utils.separator}comments${this.utils.separator}${documentName}`;
                    const docFullPathname = `${this.fsa.getBasePath()}${this.utils.separator}${documentNameAndPath}`;

                    this.sourceAndDocumentLink.add(documentNameAndPath);

                    this.doActualScanning(documentNameAndPath, srcFile, docFullPathname);
                })
                this.phaseCount = 2.0;
            }
        }

        if (this.phaseCount == 4.0 && this.testFileListWithExtension.length > 0) {
            const filesToCheck = this.testFileListWithExtension.pop();
            if (filesToCheck != undefined) {
                filesToCheck.forEach(srcFile => {

                    const testDocumentName = this.createDocNameFromSourceName(srcFile, this.testPath);
                    const testDocumentNameAndPath =
                        `${this.documentPath}${this.utils.separator}test comments${this.utils.separator}${testDocumentName}`;
                    const docFullPathname = `${this.fsa.getBasePath()}${this.utils.separator}${testDocumentNameAndPath}`;

                    this.doActualScanning(testDocumentNameAndPath, srcFile, docFullPathname);
                })
                this.phaseCount = 4.0;
            }
        }

        if (this.phaseCount == 5.0) {
            //
            // every md document that does not have a source file must be removed
            // the source and document link is the source file that should be in the 
            // document file list. 
            //            
            this.documentFileListWithExtension.forEach(fileName => {
                if (!this.sourceAndDocumentLink.has(fileName)) {
                    this.utils.fsa.remove(fileName)
                }
            })

            this.sourceAndDocumentLink.clear();
            this.phaseCount = -1.0;
        }

        this.phaseCount += 1;
    }

    /**
     * Given the file system full doc name and path scan the source file and place document 
     * in the document name and path location
     * @param documentNameAndPath vault path of the md document
     * @param srcFile to scan source file to scan
     * @param docFullPathname fila document name and path from the root of the file system
     */
    private doActualScanning(documentNameAndPath: string, srcFile: string, docFullPathname: string) {
        this.sourceAndDocumentLink.add(documentNameAndPath);

        const srcFileExists = existsSync(srcFile);
        if (!srcFileExists) {
            console.info('Test source file gone ' + srcFile);
        } else {
            const srcStat = statSync(srcFile);

            var createdFile = false;
            const docFileExists = existsSync(docFullPathname);
            var docStat: any;
            var createdFile = false;
            if (docFileExists) {
                docStat = statSync(docFullPathname);
            } else {
                writeFileSync(docFullPathname, "");
                createdFile = true;
            }
            docStat = statSync(docFullPathname);
            //
            // source is older than the doc as seen from 1970 -> onwards OR
            // docs have just been created.
            // 
            if (createdFile || docStat.mtimeMs < srcStat.mtimeMs) {
                const srcLines = readFileSync(srcFile, { encoding: 'utf8', flag: 'r' });
                var allComments;
                var comments = "NONE";
                try {
                    comments = this.codeScanner(srcLines);
                    if (comments != undefined) {
                        allComments = comments.replaceAll(/\n\s+\*/g, "\n")
                    }
                } catch (exception) {
                    console.log("Error in scan for file " + srcFile); 
                    const headerComment = `[Source](file://${srcFile})\n\n---\n`;
                    this.fsa.write(documentNameAndPath, headerComment + comments);
                }
                if (allComments == "unpaired surrogates") {
                    console.log("Error in scan for file " + srcFile); 
                }
                const headerComment = `[Source](file://${srcFile})\n\n---\n`;
                this.fsa.write(documentNameAndPath, headerComment + allComments);
            }
        }
    }

    /**
     * Create a document file name using the source file name
     * @param sourceFile to create a document file from
     * @returns the document file
     */
    createDocNameFromSourceName(sourceFile: string, applicationPath: string) {
        // get only the name of the file
        var fileName = sourceFile
            .replace(applicationPath + this.utils.separator, '')
            .replace(this.codeExtension, '.md');
        // replace all accurences
        while (fileName.contains(this.utils.separator)) {
            fileName = fileName.replace(this.utils.separator, '.');
        }
        return fileName;
    }

    /**
     * create a path to the source file relative to the vault
     * @param sourceFile the source code file
     * @param documentPart the document file
     * @returns the path to the source relative to the vault
     */
    createRelativePath(sourceFile: string, documentPart: string) {
        var sourceFileParts = sourceFile.split(this.utils.separator);
        var documentAndPathParts = `${this.utils.fsa.getBasePath()}/${documentPart}`.split('/');

        while (sourceFileParts[0] == documentAndPathParts[0]) {
            sourceFileParts = sourceFileParts.slice(1);
            documentAndPathParts = documentAndPathParts.slice(1);
        }
        // go up from current folder documentAndPathParts length - 1
        return documentAndPathParts.filter(value => {
            return !value.endsWith('.md')
        })
            .map(value => '..')
            .concat(sourceFileParts).join('/');


    }
}